### DISCUSSION TITLE
Add compatibility with cart transform functions

### DISCUSSION THREAD

**Discussion by a47ae** [Shopify/function-examples#434] [Feedback: Discounts APIs]
Hey, I'm referencing a discussion with Nick Wesselman, indicating that Shopify's current architecture doesn’t allow discount functions to interact with individual components of a bundled item (https://community.shopify.com/c/shopify-functions/combine-discount-function-and-cart-transformation/td-p/2409718).

As it stands, cart transformation functions run before discount functions and discount functions can not access item components. This limitation hinders our ability to apply discounts to individual bundle components, a feature particularly crucial for our use case.

**Use Case Context**:
We manage a loyalty program where select customers receive specific items at no cost (e.g., a complimentary engraving), facilitated through discount functions. Concurrently, the shop utilizes cart transformations to aggregate products, presenting them as a single bundled item during checkout. The issue arises when an eligible 'free' item is embedded within a bundle, rendering our discount function ineffective in adjusting the item’s price within that bundle.

Using cart transformation for price adjustments is not a viable workaround due to:

1. The inability to track these adjustments as 'discounts,' which we rely on to keep track of used discounts.
2. The potential complexity and inefficiency of intertwining business logic for discount application and cart transformation. This would also be impossible if not both functions are from the same supplier.

**Request**:
We propose enhancing the discount function framework to allow visibility and operability on individual bundled item components. Alternatively, a sequence alteration enabling item-specific discount functions to precede cart transformations could address this issue.

**Comment by gavin667** [comment]
Hey @a47ae I was just wondering if you ever received any feedback on the above or managed to work around the problem at all? We are having exactly the same issue. Thanks!

**Comment by ccarlosm** [comment]
Yep, this feature would be incredibly useful for managing discounts within bundles. The current limitation makes it difficult to apply item-specific discounts when using cart transformations.

Has there been any update from Shopify on this? Would love to know if there are any planned improvements or possible workarounds.

**Comment by oribenez** [comment]
We are also looking forward for a solution for this situation.

**Comment by sergei-sss** [comment]
It's quite strange to see a description like this:

> Functions execute in a specific sequence during checkout, and each function depends on data from earlier steps. This sequencing is important because:
> Your function's input data comes from previous checkout operations.

(from the guide: https://shopify.dev/docs/api/functions/latest#function-execution-order-in-checkout)

and still receive the same wrapper product in the discount function as in the cart transform input, instead of the expanded products after the cart transform expand operation.
